### PR TITLE
disk_partitioner base classes

### PR DESCRIPTION
**Pull request by xek**
Each kind of setup operation is an independent class.
This also adds examples and basic tests for disk_config validation

### CODE REVIEW

File: ironic-lib/0002-Base-classes-and-examples.patch [hunk 1]
```diff
@@ -0,0 +1,2283 @@
...
++# See the License for the specific language governing permissions and
++# limitations under the License.
++
++from ironic_lib.system_installer.base import Setup
++
++
++class MegaCliSetup(Setup):
++    """Hardware RAID setup implementation"""
++
++    conf_key = 'megacli'
```
Comment (jimrollenhagen) [inline ironic-lib/0002-Base-classes-and-examples.patch]: I assume the rest is implemented in a future patch?

File: ironic-lib/0002-Base-classes-and-examples.patch [hunk 2]
```diff
@@ -0,0 +1,2283 @@
...
++        def construct_mapping(loader, node):
++            loader.flatten_mapping(node)
++            return collections.OrderedDict(loader.construct_pairs(node))
++        OrderedLoader.add_constructor(
++            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
++            construct_mapping)
++        return yaml.load(disk_conf, OrderedLoader)
++
++
++def retry(tries, exc=(Exception,)):
```
Comment (jimrollenhagen) [inline ironic-lib/0002-Base-classes-and-examples.patch]: Might be able to just use this instead? https://docs.openstack.org/oslo.service/ocata/api/loopingcall.html#oslo_service.loopingcall.RetryDecorator

**Review by jimrollenhagen — COMMENTED** [review]
Looks pretty good overall, though hard to tell until I read the patches on top that use this. :)

Just a couple comments inline.

File: ironic-lib/0002-Base-classes-and-examples.patch (hunk 2 above)
Comment (xek) [reply ironic-lib/0002-Base-classes-and-examples.patch]: @PiotrProkop ^

File: ironic-lib/0002-Base-classes-and-examples.patch (hunk 1 above)
Comment (xek) [reply ironic-lib/0002-Base-classes-and-examples.patch]: Yes, in https://github.com/yahoo/openstack/pull/13, generally we implemented each of those classes/functionalities in separate PRs

File: ironic-lib/0002-Base-classes-and-examples.patch (hunk 2 above)
Comment (PiotrProkop) [reply ironic-lib/0002-Base-classes-and-examples.patch]: sure didn't notice that 😄

File: ironic-lib/0002-Base-classes-and-examples.patch (hunk 1 above)
Comment (jimrollenhagen) [reply ironic-lib/0002-Base-classes-and-examples.patch]: :+1:

**Comment by epim** [comment]
In terms of the spec and overall approach, this looks good to me. I'll leave it to Jim and Arun to discuss the finer details of the code

**Comment by jimrollenhagen** [comment]
Merging this, not fully tested (waiting for hardware) but will clean up later as needed.

**Pull request merged**